### PR TITLE
backup: move file sst sink to separate package

### PR DESCRIPTION
**Pull request by kev-cao**
This patch moves `file_sst_sink` into its own package and exports a sink writer interface to facilitate writing to the sink. Doing so grants the flexibility to provide multiple methods of writing to the sink.

Epic: none

Release note: None

### CODE REVIEW

File: pkg/backup/backupsink/span_writer.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	gogotypes "github.com/gogo/protobuf/types"
+)
+
+type FileSSTSinkExportedSpanWriter struct {
+	fileSSTSink
```
Comment (kev-cao) [inline pkg/backup/backupsink/span_writer.go:28]: Thoughts on embedding the struct here vs having a dedicated field for it?

**Review by msbutler — COMMENTED** [review]
Could you restrict this PR's scope to only moving the file sink to its own package and capitalizing the 3 methods on fileSink that need to be exported? 

I think creating the FileSSTSinkWriter should be done in a separate PR-- further, before we do this, i'd like david to weigh in.

Moving the file sink to its own package is uncontroversial though.

File: pkg/backup/backupsink/sink_utils.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/builtins"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/errors"
+)
+
+// ElidedPrefix returns the prefix of the key that is elided by the given mode.
+func ElidedPrefix(key roachpb.Key, mode execinfrapb.ElidePrefix) ([]byte, error) {
```
Comment (msbutler) [inline pkg/backup/backupsink/sink_utils.go:28]: ElidePrefix doesn't really belong in file sink as we use this func during restore, where no file sink is involved. If the key rewriter was in its own package, which it really should be (we use it in pcr and ldr), this ElidedPrefix util would live there. Fine to keep this as is, but i'm making a mental note for myself to move the key rewriter into its own package once this merges.

File: pkg/backup/backupsink/sink_utils.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	// still above any key in the file, the end key likely came from export's
+	// iteration stopping early and setting the end to the resume key, i.e. the
+	// next real family key. In this case, we are not mid-row, but want to adjust
+	// our span end -- and where we resume the next file -- to be this row key.
+	// Thus return the truncated row key and false.
+	return endRowKey, false
+
+}
+
+func (s *FileSSTSink) copyPointKeys(ctx context.Context, dataSST []byte) (roachpb.Key, error) {
```
Comment (msbutler) [inline pkg/backup/backupsink/sink_utils.go:88]: why do these two methods that hang off FileSSTSink live in sink_utils? To make this pr simpler, perhaps it is worth just throwing all this stuff in file_sst_sink.go . Github may even detect that this is simple move, like it did for `file_sst_sink_test.go`

Comment (kev-cao) [reply pkg/backup/backupsink/sink_utils.go:88]: Oh at the time of refactoring I thought I made throw in some more helper functions in here later down the line for backup compactions. Although thinking about it, these are specific to the file sink implementation. Will move it back

**Review by msbutler — APPROVED** [review]

**Comment by kev-cao** [comment]
TFTR!

bors r=msbutler

**Pull request merged**